### ISSUE TITLE
Reliable way to debug Nan in backward pass?

### ISSUE THREAD

**Issue by charleswilmot** [patrick-kidger/diffrax#742]
Hi diffrax!

I am reliably poisoning my gradient with nans, but I have not yet figured **how to reliably bisect the source of the poisoning**.

I read everything about [gradients containing NaN where using `where`](https://docs.jax.dev/en/latest/faq.html#gradients-contain-nan-where-using-where), as well as [this detailed answer from mattjj](https://github.com/jax-ml/jax/issues/1052#issuecomment-514083352). I checked my square roots too (looking for `sqrt'(0) = inf`), but this feels a lot like guess work.

I think a better approach would be to use `jax.config.update("jax_debug_nans", True)` or maybe to use checkify on the differentiated function.

Do you have any recommendations?

EDIT: so it turns out that in my case, I just needed to use a smaller dt0 and now the gradient is healthy. I would still be interested in a reliable way to back-trace where gradient poisoning originates from.

**Comment by patrick-kidger** [comment]
I'd recommend `equinox.debug.backward_nan` :)

**Comment by charleswilmot** [comment]
I just found out about #280 , where you gave a very complete answer already.

Anyways, thanks for answering the question again! I am closing this issue :D
